### ISSUE TITLE
Return type notation with constant parameter can potentially cause misleading diagnostics 

### ISSUE THREAD

**Issue by c410-f3r** [rust-lang/rust#127776]
`RTN` currently does not support constant parameters. However, when they are present, it prints the same diagnostics that hint potential bounds of a successful compilation.

```rust
#![allow(incomplete_features)]
#![feature(return_type_notation)]

use core::future::Future;

fn enforce_send<T: Send>(_: T) {}

trait Foo {
    fn foo(&self, _n: [(); 4]) -> impl Future {
        async {}
    }
}

fn stuff<F>(foo: F)
where
    F: Foo,
    for<'any> &'any F: Send
{
    enforce_send(async {
       foo.foo([(); 4]).await; 
    });
}
```

> help: within `{async block@src/lib.rs:20:16: 20:21}`, the trait `Send` is not implemented for `impl Future { <F as Foo>::foo() }`, which is required by `{async block@src/lib.rs:20:16: 20:21}: Send`

Users see the above hint and can quickly figure out the `F: Foo<foo(..): Send>` missing piece. Unfortunately that is not true when dealing with constant parameters.

```rust
#![allow(incomplete_features)]
#![feature(return_type_notation)]

use core::future::Future;

fn enforce_send<T: Send>(_: T) {}

trait Foo {
    fn foo<const N: usize>(&self, _n: [(); N]) -> impl Future {
        async {}
    }
}

fn stuff<F>(foo: F)
where
    F: Foo,
    for<'any> &'any F: Send
{
    enforce_send(async {
       foo.foo([(); 4]).await; 
    });
}
```

> help: within `{async block@src/lib.rs:20:16: 20:21}`, the trait `Send` is not implemented for `impl Future { <F as Foo>::foo<4>() }`, which is required by `{async block@src/lib.rs:20:16: 20:21}: Send`

* If `F: Foo(foo(..): Send`, constant parameters are unsupported. 
* If `F: Foo(foo<4>(..): Send`, syntax error.
* If `F: Foo(foo::<4>(..): Send`, syntax error.

cc @compiler-errors in case you are interested.

**Comment by c410-f3r** [comment]
The `impl Future { <F as Foo>::foo<4>() }` part that could lead users to try things like `F: Foo(foo::<4>(..): Send` is no longer appearing in the latest `rustc` version. Therefore, I will close this issue.
